hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsHttpOperation.java

```diff
@@ -314,13 +327,46 @@ public void sendRequest(byte[] buffer, int offset, int length) throws IOExceptio
     if (this.isTraceEnabled) {
       startTime = System.nanoTime();
     }
-    try (OutputStream outputStream = this.connection.getOutputStream()) {
-      // update bytes sent before they are sent so we may observe
-      // attempted sends as well as successful sends via the
-      // accompanying statusCode
+    OutputStream outputStream = null;
+    // Updates the expected bytes to be sent based on length.
+    this.expectedBytesToBeSent = length;
+    try {
+      try {
+        /* Without expect header enabled, if getOutputStream() throws
+           an exception, it gets caught by the restOperation. But with
+           expect header enabled we return back without throwing an exception
+           for the correct response code processing.
+         */
+        outputStream = this.connection.getOutputStream();
+      } catch (IOException e) {
+        /* If getOutputStream fails with an exception and expect header
+           is enabled, we return back without throwing an exception to
+           the caller. The caller is responsible for setting the correct status code.
+           If expect header is not enabled, we throw back the exception.
+         */
+        String expectHeader = this.connection.getRequestProperty(EXPECT);
+        if (expectHeader != null && expectHeader.equals(HUNDRED_CONTINUE)) {
+          LOG.debug("Getting output stream failed with expect header enabled, returning back "
+                  + ExceptionUtils.getStackTrace(e));
+          return;
```

Comment:
The request was sent to the server with expect header enabled but it didn't respond back with 100-continue because client didn't meet with the expectations. So the connection was successfully set and hence it can set all these properties accordingly.

Reply:
Are you saying that server in this case would send some status with >400, and that would be caught in https://github.com/anmolanmol1234/hadoop/blob/HADOOP-18146/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsHttpOperation.java#L412-L416?

If yes, what if getOutputStream fails due to openSSL handshake like:
```
java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.wildfly.openssl.OpenSSLSocket.runHandshake(OpenSSLSocket.java:306)
	at org.wildfly.openssl.OpenSSLSocket.startHandshake(OpenSSLSocket.java:210)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:197)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1342)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1317)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:264)
	at org.apache.hadoop.fs.azurebfs.services.AbfsHttpOperation.sendRequest(AbfsHttpOperation.java:300)
```

Also, it would be stretch of logic accrosss multiple places: in sendRequest(), processresponse(), abfsclient.append(). It would be awesome if we flow an exception from this point to abfsclient. What you feel?

Reply:
So, the response code will depend on getOutputStream failed due to which reason, if it fails due to openSSL Handshake and expect header is enabled, it will get the status of 0 or -1 in the processResponse being called in AbfsHttpOperation and thus the request will get retried.  The status code of >400 is not guaranteed, the correct understanding of what the response code is can be done by the getResponseCode() in the processResponse only. So would not be right to throw and exception from this point. The flow should be from AbfsRestOperation to client only.